File: almostunionfind.py
from sys import stdin,stdout

class unionFind:
    def __init__(self,N):
        self.p2 = [i for i in range(N)] #keeps track of original parent when a move is made
        self.p = [i for i in range(N)]
        self.rank = [0 for i in range(N)]
        self.numElems = [1 for i in range(N)]
        self.total = [i for i in range(N)]
    '''
    def debug(self):
        print(f'p2 = {self.p2}')
        print(f'p = {self.p}')
        print(f'numElems = {self.numElems}')
        print(f'total = {self.total}')
    ''' 
    def findSet(self,j):
        a = self.p2[j]
        b = self.p[a]
        while a != b:
            self.p[a] = self.p[b]
            a = self.p[a]
            b = self.p[a]
        return a
        
    def isSameSet(self,i,j):
        return self.findSet(i) == self.findSet(j)

    def unionSet(self,i,j):
        if not self.isSameSet(i,j):
            x,y = self.findSet(i),self.findSet(j)
            # union by rank
            if self.rank[x] > self.rank[y]:
                x,y = y,x
            self.p[y] = x
            self.numElems[x] += self.numElems[y]
            self.numElems[y] = 0
            self.total[x] += self.total[y]
            self.total[y] = 0
        
    def move(self,a,b): 
        if not self.isSameSet(a,b):
            x,y = self.findSet(a),self.findSet(b)
            self.numElems[y] += 1
            self.numElems[x] -= 1
            self.total[y] += a
            self.total[x] -= a
            self.p2[a] = y
        
    def numberOfElems(self,i):
        return self.numElems[self.findSet(i)]

    def elemSum(self,i): # add back due to zero indexing
        return self.total[self.findSet(i)] + self.numberOfElems(i)
                            
while True:
    try:
        n,m = map(int,stdin.readline().split())
        UFDS = unionFind(n)
        for i in range(m):
            command = stdin.readline().split()
            if command[0] == '1': #-1 to offset for zero indexing
                UFDS.unionSet(int(command[1])-1,int(command[2])-1)
            elif command[0] == '2':
                UFDS.move(int(command[1])-1,int(command[2])-1)
            else: #'3' command
                stdout.write(f'{UFDS.numberOfElems(int(command[1])-1)} {UFDS.elemSum(int(command[1])-1)}\n')
    except: #EOF error
        break
    
    
    

File: almostunionfind.java
import java.io.*;

class UFDS {                                              
    public int[] p,p2,rank,numElems;
    public long[] total;

    public UFDS (int N) {
        p = new int[N]; p2 = new int[N]; rank = new int[N];
        numElems = new int[N]; total = new long[N];
        for (int i = 0; i < N; i++) {
            p[i] = i;
            p2[i] = i; //keeps track of original parent when a move command is made
            rank[i] = 0;
            numElems[i] = 1; //keeps track of number of elements in each disjoint set
            total[i] = i; //keeps track of sum of elements in each each disjoint set
        }
    }

    public int findSet(int i) { 
        int a = p2[i];
        int b = p[a];
        while (a != b) {//while parent of i and parent of parent of i do not match
            p[a] = p[b];
            a = p[a];
            b = p[a];
        }
        return a;
    }

    public Boolean isSameSet(int i, int j) {return findSet(i) == findSet(j);}

    public void unionSet(int i, int j) { 
        if (!isSameSet(i, j)) { 
            int x = findSet(i), y = findSet(j);
            // union by rank
            if (rank[x] > rank[y]){// swap x and y if rank[x] > rank[y] to shorten code 
                int temp = x;
                x = y;
                y = temp;
            }
            p[y] = x;
            numElems[x] += numElems[y];
            numElems[y] = 0; 
            total[x] += total[y];
            total[y] = 0;
        }
    }

    public void move(int i, int j){
        if (!isSameSet(i,j)){
            int x = findSet(i), y = findSet(j);
            numElems[x] -= 1; numElems[y] += 1;
            total[x] -= i; total[y] += i;
            p2[i] = y; //keep track of original parent
        }
    }

    public int numOfElements(int i) {return numElems[findSet(i)];}
    // number of elements is added back to offset zero indexing
    public long findSum(int i) {return total[findSet(i)] + numOfElements(i);}

}

public class almostunionfind {
    public static void main(String[] args) throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        PrintWriter pw = new PrintWriter(new BufferedWriter(new OutputStreamWriter(System.out)));
        while (true){
            try {
                String[] s = br.readLine().split(" ");
                int N = Integer.parseInt(s[0]), M = Integer.parseInt(s[1]);
                UFDS disjointSets = new UFDS(N);
                for (int i = 0; i < M; i++){
                    String[] command = br.readLine().split(" ");
                    if (command[0].equals("3")) {
                        int x = Integer.parseInt(command[1]) - 1; //-1 for zero indexing
                        pw.printf("%d %d\n",disjointSets.numOfElements(x),disjointSets.findSum(x));
                    } 
                    else {
                        int x = Integer.parseInt(command[1]) - 1, y = Integer.parseInt(command[2]) - 1; //-1 for zero indexing
                        if (command[0].equals("1")) {disjointSets.unionSet(x, y);}
                        else {disjointSets.move(x, y);}
                    }
                }
            }
            catch (Exception e){ // check for EOF error
                break;
            }
        }
        br.close();
        pw.close();
    }
}

